File: MTGSearch/src/main/java/com/dbottillo/mtgsearchfree/view/fragments/SavedFragment.java
package com.dbottillo.mtgsearchfree.view.fragments;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import android.widget.Toast;

import com.dbottillo.mtgsearchfree.R;
import com.dbottillo.mtgsearchfree.exceptions.MTGException;
import com.dbottillo.mtgsearchfree.model.CardsBucket;
import com.dbottillo.mtgsearchfree.model.DeckBucket;
import com.dbottillo.mtgsearchfree.model.MTGCard;
import com.dbottillo.mtgsearchfree.presenter.CardsPresenter;
import com.dbottillo.mtgsearchfree.util.LOG;
import com.dbottillo.mtgsearchfree.util.TrackingManager;
import com.dbottillo.mtgsearchfree.view.CardsView;
import com.dbottillo.mtgsearchfree.view.activities.CardsActivity;
import com.dbottillo.mtgsearchfree.view.activities.MainActivity;
import com.dbottillo.mtgsearchfree.view.adapters.CardsAdapter;
import com.dbottillo.mtgsearchfree.view.adapters.OnCardListener;
import com.dbottillo.mtgsearchfree.view.helpers.CardsHelper;
import com.dbottillo.mtgsearchfree.view.helpers.DialogHelper;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import fr.castorflex.android.smoothprogressbar.SmoothProgressBar;

public class SavedFragment extends BasicFragment implements OnCardListener, MainActivity.MainActivityListener, CardsView {

    private CardsBucket savedBucket;

    @BindView(R.id.progress)
    SmoothProgressBar progressBar;
    @BindView((R.id.empty_view))
    TextView emptyView;
    @BindView(R.id.card_list)
    RecyclerView listView;

    MainActivity mainActivity;

    @Inject
    CardsPresenter cardsPresenter;

    @Inject
    CardsHelper cardsHelper;

    public void onAttach(Context context) {
        super.onAttach(context);
        mainActivity = (MainActivity) context;
        mainActivity.setMainActivityListener(this);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.fragment_set, container, false);
        ButterKnife.bind(this, rootView);
        return rootView;
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        emptyView.setText(R.string.empty_saved);
        setActionBarTitle(getString(R.string.action_saved));

        LinearLayoutManager llm = new LinearLayoutManager(getContext());
        listView.setLayoutManager(llm);

        setHasOptionsMenu(true);

        getMTGApp().getUiGraph().inject(this);
        cardsPresenter.init(this);
    }

    @Override
    public void onResume() {
        super.onResume();
        cardsPresenter.loadFavourites();
    }

    @Override
    public String getPageTrack() {
        return "/saved";
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
        inflater.inflate(R.menu.refresh, menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int i1 = item.getItemId();
        if (i1 == R.id.action_refresh) {
            cardsPresenter.loadFavourites();
            return true;
        }
        return false;
    }

    @Override
    public void onCardSelected(MTGCard card, int position, View view) {
        LOG.d();
        TrackingManager.trackOpenCard(position);
        startActivity(CardsActivity.newFavInstance(getContext(), position));
    }

    @Override
    public void onOptionSelected(MenuItem menuItem, MTGCard card, int position) {
        LOG.d();
        if (menuItem.getItemId() == R.id.action_add_to_deck) {
            DialogHelper.open(dbActivity, "add_to_deck", AddToDeckFragment.newInstance(card));

        } else if (menuItem.getItemId() == R.id.action_remove) {
            cardsPresenter.removeFromFavourite(card, false);
            cardsPresenter.loadFavourites();
        }
    }

    @Override
    public void updateContent() {
        LOG.d();
        CardsBucket bucket = cardsHelper.filterCards(mainActivity.getCurrentFilter(), null, savedBucket);
        CardsAdapter adapter = CardsAdapter.list(bucket, false, R.menu.card_saved_option);
        adapter.setOnCardListener(this);
        listView.setAdapter(adapter);
        emptyView.setVisibility(bucket.getCards().size() == 0 ? View.VISIBLE : View.GONE);
    }

    @Override
    public void cardLoaded(CardsBucket bucket) {
        LOG.d();
        progressBar.setVisibility(View.GONE);
        savedBucket = bucket;
        updateContent();
    }

    @Override
    public void deckLoaded(DeckBucket bucket) {
        throw new UnsupportedOperationException();
    }


    @Override
    public void favIdLoaded(int[] favourites) {
        throw new UnsupportedOperationException();
    }

    @Override
    public void cardTypePreferenceChanged(boolean grid) {
        throw new UnsupportedOperationException();
    }

    @Override
    public void showError(String message) {
        LOG.d();
        Toast.makeText(getActivity(), message, Toast.LENGTH_SHORT).show();
        TrackingManager.trackSearchError(message);
    }

    @Override
    public void showError(MTGException exception) {

    }

}

File: MTGSearch/src/main/java/com/dbottillo/mtgsearchfree/ui/saved/SavedFragment.kt
package com.dbottillo.mtgsearchfree.ui.saved

import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.Toast
import com.dbottillo.mtgsearchfree.R
import com.dbottillo.mtgsearchfree.exceptions.MTGException
import com.dbottillo.mtgsearchfree.model.*
import com.dbottillo.mtgsearchfree.ui.BaseHomeFragment
import com.dbottillo.mtgsearchfree.ui.cardsConfigurator.CardsConfiguratorFragment
import com.dbottillo.mtgsearchfree.util.LOG
import com.dbottillo.mtgsearchfree.util.TrackingManager
import com.dbottillo.mtgsearchfree.ui.cards.CardsActivity
import com.dbottillo.mtgsearchfree.ui.search.SearchActivity
import com.dbottillo.mtgsearchfree.ui.cards.OnCardListener
import com.dbottillo.mtgsearchfree.ui.decks.AddToDeckFragment
import com.dbottillo.mtgsearchfree.ui.DialogHelper
import com.dbottillo.mtgsearchfree.ui.views.MTGCardsView
import com.dbottillo.mtgsearchfree.ui.views.MTGLoader
import javax.inject.Inject

class SavedFragment : BaseHomeFragment(), SavedCardsView, OnCardListener {

    lateinit var loader: MTGLoader
    lateinit var mtgCardsView: MTGCardsView
    lateinit var emptyContainer : LinearLayout

    @Inject
    lateinit var savedCardsPresenter: SavedCardsPresenter

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val rootView = inflater?.inflate(R.layout.fragment_saved, container, false)
        app.uiGraph.inject(this)
        return rootView
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        mtgCardsView = view.findViewById<MTGCardsView>(R.id.cards)
        loader = view.findViewById<MTGLoader>(R.id.loader)
        emptyContainer = view.findViewById<LinearLayout>(R.id.empty_saved_cards_container)
        view.findViewById<View>(R.id.empty_cards_action).setOnClickListener{ openSearch() }

        setupHomeActivityScroll(viewRecycle = mtgCardsView.listView)

        savedCardsPresenter.init(this)
    }

    override fun onResume() {
        super.onResume()
        savedCardsPresenter.load()
    }

    override fun onPause() {
        super.onPause()
        savedCardsPresenter.onPause()
    }

    override fun getScrollViewId(): Int {
        return R.id.card_list
    }

    override fun getPageTrack(): String {
        return "/saved"
    }

    override fun getTitle(): String {
        return context.getString(R.string.action_saved)
    }

    override fun hideLoading() {
        loader.visibility = View.GONE
    }

    override fun showLoading() {
        loader.visibility = View.VISIBLE
    }

    override fun showError(message: String?) {
        LOG.d()
        Toast.makeText(context, message, Toast.LENGTH_SHORT).show()
        TrackingManager.trackSearchError(message)
    }

    override fun showError(exception: MTGException?) {
        LOG.d()
        Toast.makeText(activity, exception?.message, Toast.LENGTH_SHORT).show()
        TrackingManager.trackSearchError(exception?.message)
    }

    override fun showCards(cardsCollection: CardsCollection) {
        emptyContainer.visibility = View.GONE
        mtgCardsView.visibility = View.VISIBLE
        mtgCardsView.loadCards(cardsCollection.list, this, getString(R.string.action_saved), cardsCollection.filter, R.menu.card_saved_option)
    }

    override fun showEmptyScreen() {
        mtgCardsView.visibility = View.GONE
        emptyContainer.visibility = View.VISIBLE
    }

    override fun showCardsGrid() {
        LOG.d()
        mtgCardsView.setGridOn()
    }

    override fun showCardsList() {
        LOG.d()
        mtgCardsView.setListOn()
    }

    override fun onCardSelected(card: MTGCard, position: Int) {
        TrackingManager.trackOpenCard(position)
        startActivity(CardsActivity.newFavInstance(context, position))
    }

    override fun onOptionSelected(menuItem: MenuItem, card: MTGCard, position: Int) {
        LOG.d()
        when(menuItem.itemId){
            R.id.action_add_to_deck -> DialogHelper.open(dbActivity, "add_to_deck", AddToDeckFragment.newInstance(card))
            R.id.action_remove -> {
                savedCardsPresenter.removeFromFavourite(card)
                savedCardsPresenter.load()
            }
        }
    }

    override fun onCardsViewTypeSelected() {
        LOG.d()
        savedCardsPresenter.toggleCardTypeViewPreference()
    }

    override fun onCardsSettingSelected() {
        val cardsConfigurator= CardsConfiguratorFragment()
        cardsConfigurator.show(dbActivity.supportFragmentManager, "cards_configurator")
        cardsConfigurator.listener = object : CardsConfiguratorFragment.CardsConfiguratorListener{
            override fun onConfigurationChange() {
                savedCardsPresenter.load()
            }
        }
    }

    fun openSearch(){
        LOG.d()
        startActivity(Intent(activity, SearchActivity::class.java))
    }

    override fun onTitleHeaderSelected() {

    }

}
